### PR TITLE
Optional chaining for cell paths (`?.`)

### PR DESCRIPTION
**Pull request by rgwood**
This PR adds the ability to use `?.` in cell paths, somewhat like [optional chaining in JS](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) and [C#'s null-conditional operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-).

This provides a succinct way for users to specify whether a failed cell path access should return a `ShellError` or `Value::Nothing`. It is not aiming for exact compatibility with JS and C#.

### Examples

```bash
# fails because Value::Nothing has no data to access
$nothing.foo
# returns Value::Nothing because `?` was used
$nothing?.foo

# fails because `asdf` is not a column on the record
{foo: 'bar'}.asdf
# returns Value::Nothing because `?` was used
{foo: 'bar'}?.asdf

# fails because `asdf` is not a column on the nested record
{foo: {bar: 'baz'} }.foo.asdf
# returns Value::Nothing
{foo: {bar: 'baz'} }.foo?.asdf

# fails because there is no 10th row in the list
[{foo: 'bar'}].10
# returns Value::Nothing
[{foo: 'bar'}]?.10

# fails because there is no `foo` column in the 2nd row
# BREAKING CHANGE - PREVIOUSLY THIS WOULD SUCCEED AS LONG AS 1+ ROW HAD A `foo` COLUMN
[{foo: 'bar'} {}].foo

# returns a `list<any>` with 2 values: "bar" and $nothing
[{foo: 'bar'} {}]?.foo
```

### How does this relate to `get` and `select`?

As always, any cell path access can be rewritten to use `get` (ex: `$foo.a` -> `$foo | get a`). The arguments given to `get` and `select` (ex: `foo` and `bar` in `select foo bar`) are actually cell paths.

Before this PR, cell paths used by `get` and `select` were not allowed to start with a `.`; parsing `get .a` would fail. After this PR, a `.` before the first member is (optionally) allowed. All of the following are now valid:

```bash
get a
get .a
get ?a
get ?.a
```

I think this makes sense; the user's intent is unambiguous in each case and it provides consistency whether using `get` or accessing a cell path without `get`.

As part of this PR, the `-i`/`--ignore-errors` flag has been removed from `get` and `select`. Using `?` in cell paths now offers more fine-grained control over error handling:

```bash
# this fails because `b` is missing on the 2nd row (not new behaviour)
[{a: 1, b: 2} {a: 3}] | select b

# previously you had to use something like `default` to convert missing values to null
[{a: 1, b: 2} {a: 3}] | default null b | select b

# the old `-i` flag would return a single null value instead of a table with some null cells - usually not what you want!
[{a: 1, b: 2} {a: 3}] | select -i b

# now you can use `?`
[{a: 1, b: 2} {a: 3}] | select ?b
```

### Future Work

If this change lands, I should be able to follow it up with another PR that makes cell path access stream properly instead of collecting `ListStream`s.

### CODE REVIEW

**Comment by rgwood** [comment]
cc: @webbedspace

File: src/tests/test_cell_paths.rs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+#[test]
+fn list_row_optional_access_succeeds() -> TestResult {
+    run_test("[{foo: 'bar'}, {foo: 'baz'}]?.2", "")
+}
+
+// tests for ListStreams, currently unused
+// TODO: re-enable these tests once we have a way to do streaming cell path access
+// I'm using a hack to create ListStreams for testing: pipe a list into `each`
+
+// #[test]
```
[the remark this answers is not in the document]
Comment (rgwood) [reply src/tests/test_cell_paths.rs]: Good call, done.

File: crates/nu-command/src/filters/get.rs [hunk 2]
```diff
@@ -114,6 +92,14 @@ impl Command for Get {
                     span: Span::test_data(),
                 }),
             },
+            Example {
+                description: "Get a column from a table, and return null for any rows where that column does not exist",
+                example: "[{A: A0} {}] | get ?.A",
```
Comment (webbedspace) [inline crates/nu-command/src/filters/get.rs:97]: Consider adding an example for `get .A` as well, with the text indicating that the dot is strictly optional. Such an example would provide needed clarity as to what the `?.` syntax "means". (Currently, it sorta looks like the ? is another cell name, when actually it's an extension of the separator between names.)

**Comment by webbedspace** [comment]
Could you add tests that this works with `into cellpath`? (i.e. Having ?. in a "cellpath string" works correctly.)

**Comment by rgwood** [comment]
@webbedspace `into cellpath` has been deleted for now: https://github.com/nushell/nushell/pull/7523

File: src/tests/test_cell_paths.rs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+#[test]
+fn record_with_nested_list_column_failure() -> TestResult {
+    fail_test("{foo: [{bar: 'baz'}]}.foo.0.asdf", "")
+}
+
+#[test]
+fn deeply_nested_optional_cell_path() -> TestResult {
+    run_test(
+        "{foo: [{bar: 'baz'}]}.foo?.3?.bar?.asdfdafg?.234?.foobar",
```
[the remark this answers is not in the document]
Comment (rgwood) [reply src/tests/test_cell_paths.rs]: Good idea, done.

File: crates/nu-command/src/filters/get.rs (hunk 2 above)
Comment (rgwood) [reply crates/nu-command/src/filters/get.rs:97]: First attempt:
image

I feel like that could still be a little more clear. It's late here, will sleep on it.

**Comment by webbedspace** [comment]
Ah, very well then.

**Comment by rgwood** [comment]
I've confirmed that this works as expected with `select`, and removed the `-i` flag from `select`. I believe this would solve a real usability problem for `select`: https://github.com/nushell/nushell/issues/7558

**Comment by rgwood** [comment]
This proposal confused a few people on Discord. It's possible I didn't explain it well there, but it's also possible there might be better syntax out there for this kind of functionality.

I feel that this PR is ready to go from a code quality perspective, but I'd like to hear a few more people weigh in on whether they think the syntax is OK (with alternative examples if you don't like the current syntax).

**Comment by webbedspace** [comment]
Actually, now that you mention it, I just noticed this does have some peculiarities I don't know if I like. (Sorry for not scrutinising that closely earlier).

>It is not aiming for exact compatibility with JS and C#.

So, like, in JS and C#, `?.` means "return null if the left side is null". (JS uses undefined, but w/e)
Here, it actually means "return null if the right side is not in the left side", which I just realised is a little problematic for two reasons.
One, it requires the reader to read in both directions, versus the JS/C# left-to-right reading approach.
Two, because "the right side" isn't really well-defined.
Consider `$a?.b.c.d`. Is the "right side" `b.c.d` or just `b`? If `b` is `{c:{}}`, then will `$a?.b.c` work? I assume that in this case the right side is just `b`, but there is this ambiguity there.

How about this alternative:
```
# fails because Value::Nothing has no data to access
$nothing.foo
# returns Value::Nothing because `?` was used
$nothing?.foo

# fails because `asdf` is not a column on the record
{foo: 'bar'}.asdf
# returns Value::Nothing because `?` was used
{foo: 'bar'}.asdf?

# fails because `asdf` is not a column on the nested record
{foo: {bar: 'baz'} }.foo.asdf
# returns Value::Nothing
{foo: {bar: 'baz'} }.foo.asdf?
# returns 'baz'
{foo: {bar: 'baz'} }.foo?.bar
{foo: {bar: 'baz'} }.foo?.bar?
{foo: {bar: 'baz'} }?.foo?.bar

# fails
{foo: 1} | get .bar
{foo: 1} | get ?.bar
null | get .bar
null | get .bar?
# returns Value::Nothing
{foo: 1} | get ?.bar
{foo: 1} | get .bar?
{foo: 1} | get ?.bar?
null | get ?.bar
null | get ?.bar?
```
Here, I've split `?` away from the `?.`. It's almost the same, but the new "terminal `?`" makes it fully unambiguous which access is optional.

**Comment by rgwood** [comment]
OK, yeah, people on Discord had similar issues with the placing of the `?` - I didn't quite understand them at the time but if this many people have expressed concerns there's something wrong with the current approach...

> Consider `$a?.b.c.d`. Is the "right side" `b.c.d` or just `b`? If `b` is `{c:{}}`, then will `$a?.b.c` work? I assume that in this case the right side is just `b`, but there is this ambiguity there.

Correct, the `?` only applies to `b`.

I'm not sure I understand all of your examples. You mention a "terminal `?`" but it seems like you're putting it on either side of the path member. For example, for these examples that return `Value::Nothing` why is the `?` sometimes on the left and sometimes on the right?

```
# returns Value::Nothing because `?` was used
$nothing?.foo
{foo: 'bar'}.asdf?
{foo: 1} | get ?.bar
{foo: 1} | get .bar?
```

**Comment by rgwood** [comment]
My stab at the "terminal `?`" approach looks like this:

```
# fails because Value::Nothing has no data to access
$nothing.foo
# returns Value::Nothing because `?` makes the `foo` path member optional
$nothing.foo?

# fails because `asdf` is not a column on the record
{foo: 'bar'}.asdf
# returns Value::Nothing because `?` makes the `asdf` path member optional
{foo: 'bar'}.asdf?

# fails
{foo: {bar: 'baz'} }.foo.asdf
{foo: {bar: 'baz'} }.asdf.asdf
# returns Value::Nothing
{foo: {bar: 'baz'} }.foo.asdf?
{foo: {bar: 'baz'} }.asdf?.asdf?

# fails
null | get bar
{foo: 1} | get bar
# returns Value::Nothing
null | get bar?
{foo: 1} | get bar?
```

**Comment by webbedspace** [comment]
I'll explain further.
```
$nothing?.foo # if $nothing exists then $nothing.foo else null

{foo: 'bar'}.asdf? # if {foo: 'bar'}.asdf exists then {foo: 'bar'}.asdf else null

{foo: {bar: 'baz'} }.foo.asdf? # if {foo: {bar: 'baz'} }.foo.asdf exists then {foo: {bar: 'baz'} }.foo.asdf else null
{foo: {bar: 'baz'} }.foo?.bar # if {foo: {bar: 'baz'} }.foo exists then {foo: {bar: 'baz'} }.foo.bar else null
{foo: {bar: 'baz'} }.foo?.bar? # if {foo: {bar: 'baz'} }.foo exists and {foo: {bar: 'baz'} }.foo.bar exists then {foo: {bar: 'baz'} }.foo.bar else null
{foo: {bar: 'baz'} }?.foo?.bar # if {foo: {bar: 'baz'} } exists and {foo: {bar: 'baz'} }.foo exists and {foo: {bar: 'baz'} }.foo.bar exists then {foo: {bar: 'baz'} }.foo.bar else null

**Comment by rgwood** [comment]
Thanks; I don't find that intuitive right now but maybe it'll grow on me.

I think you accidentally listed `{foo: 1} | get ?.bar` twice in [your first example](https://github.com/nushell/nushell/pull/7540#issuecomment-1364495748) - could you confirm how that behaves (I'm guessing it fails)?

**Comment by webbedspace** [comment]
`{foo: 1} | get ?.bar` means "if the input (also called $in) exists then $in.bar else null"

**Comment by rgwood** [comment]
Gotcha. Might want to update your first example:
```
...
# returns Value::Nothing
{foo: 1} | get ?.bar        <----- think this line is wrong
{foo: 1} | get .bar?
...
```

**Comment by webbedspace** [comment]
Ah yes, good point, sorry.

**Comment by rgwood** [comment]
Maybe it would help to recap the motivating example for this: using `get` and `select` on lists where records don't all have the same columns.

### Current Behaviour

| Command | Result |
|---|---|
| `[{a: 1, b: 'foo'} {a: 2}] \| select b`  | failure |
| `[{a: 1, b: 'foo'} {a: 2}] \| get b`  | `['foo', null]`  |
| `[{a: 1, b: 'foo'} {a: 2}] \| get c`  | failure  |
| `[{a: 1, b: 'foo'} {a: 2}] \| select -i b`  |  `null` |
| `[{a: 1, b: 'foo'} {a: 2}] \| get -i c`  |  `null` |

Note that:
1. `get` and `select` are inconsistent in how they handle missing data
2. `get` only succeeds if it can get a value from at least 1 row; there is no way for a user to say "replace missing data with `null`, even if all rows are missing data"
3. `-i` replaces the _entire_ result with `null` - not just the missing data[^1].

### Desired Behaviour

It would be a lot simpler if there were only 2 ways to handle a failed cell path access: fail, or replace missing data with `null`. The default should probably be failure, and users should have an easy way to switch to the more lenient behaviour.

### Next Steps

I'm still not sure what the syntax should be. I think I am trying to solve a problem that is slightly different than the one JS solves with `?.`, and that has been a major point of confusion for everyone. It's currently pretty late on Christmas Eve, I'll come back to this another time.

[^1]: I think we need to get rid of `-i`. The biggest reason: `-i`'s behaviour is not compatible with streaming data. You can't change your mind halfway through a ListStream and say "actually I want to return a single `null` instead of multiple values."

**Comment by webbedspace** [comment]
>I think I am trying to solve a problem that is slightly different than the one JS solves with `?.`

The problems are very similar, however. I don't think the difference is that major.

In JS, 1) access of a missing property produces `undefined` and 2) access of a property on `undefined` is an error; `?.` is used to avoid this by having a multi-level path short-circuit.

Nushell is more strict than JS: 1) access of a missing property is an error, and 2) access of a property on `null` is an error. `?.` can be used to make a multi-level path turn to null + short-circuit, and the terminal `?` (needed for point 1) can make a single-level path turn to null.

**Comment by rgwood** [comment]
I think this is the main thing that bothers me about your proposal:

```
# returns null
$nothing?.foo
# also returns null
{foo: 'bar'}.asdf?
```

The position of the `?` moves around even though in both cases I want the same behaviour (try to get a field, return `null` if the access fails).

**Comment by webbedspace** [comment]
Well, those are not really "the same behaviour" insofar as the access fails for two different reasons in both. Accessing from `null` or a non-structure type is a different kind of error from accessing a missing key (in JS, only the first produces a proper error).

**Comment by rgwood** [comment]
> Accessing from null or a non-structure type is a different kind of error from accessing a missing key

But does the user care about that difference when they're writing a cell path? I don't _think_ they do; if I'm writing `<some list> | get foo` I want a switch that controls whether access fails and I don't especially care whether it failed because a row is `null` or a record without a `foo` column.

### Implementation

It might help me understand your point of view if you can show me roughly how you'd implement your desired solution. My proposal involves adding an `optional` field to `CellPath` like:
#### Old:
```rust
pub struct CellPath {
    pub members: Vec<PathMember>,
}
pub enum PathMember {
    String { val: String, span: Span },
    Int { val: usize, span: Span },
}
```
#### New:
```rust
pub struct CellPath {
    pub members: Vec<PathMember>,
}
pub enum PathMember {
    String { val: String, span: Span, optional: bool },
    Int { val: usize, span: Span, optional: bool },
}
```

Under your proposal, what would `CellPath` and/or `PathMember` look like?

**Comment by webbedspace** [comment]
>But does the user care about that difference when they're writing a cell path? I don't *think* they do;

Isn't the whole point of having a type system like Nushell's is that the user cares about the types of the data? Besides, if you're writing `<some list> | get foo` then it's already established that it's a list and not null, right? So the only possible error is for `foo` not being present, and thus the obvious case here is `| get foo?`.

I suppose (and I should emphasise this is fresh off the dome, and I haven't consulted the relevant code yet) my solution would be to add `optional` to both PathMember and CellPath:
* an optional PathMember only represents a member with a terminating `?` like `.foo?` or `get foo?`, which may be followed with more PathMembers to produce the `?.` formulation (`$a.foo?`, `$a.foo?.bar`, `$a.foo?.bar?`) It turns the PathMember into `null` if it isn't present on the structure, and stops evaluation of the rest of the members.
* an optional CellPath only represents an initial `?.`: `get ?.foo`, `$in?.foo`. It turns the entire CellPath into `null` if the variable (or, for `get`, the input) isn't a structure (list or record).

**Comment by rgwood** [comment]
> Besides, if you're writing `<some list> | get foo` then it's already established that it's a list and not null, right? So the only possible error is for foo not being present, and thus the obvious case here is `| get foo?`.

`get foo` on a list essentially applies the `foo` cell path to every element of the list, and those elements can be null. e.g. `[{foo: 'bar'} null]` vs `[{foo: 'bar'} {}]`

**Comment by webbedspace** [comment]
>`get foo` on a list essentially applies the foo cell path to every element of the list, and those elements can be null. e.g. `[{foo: 'bar'} null]` vs `[{foo: 'bar'} {}]`

Well, I mean, if you're doing `get foo` on a list in general, then you should have assurance that it's a proper table, right? A table is a list with records for every element, so `| get foo?` would still always work. If you really don't know this for sure, then you can do `| get ?.foo?`

**Comment by rgwood** [comment]
> Well, I mean, if you're doing get foo on a list in general, then you should have assurance that it's a proper table, right?

Yeah, that's usually a safe assumption to make - something's probably gone wrong if you have mixed records and null values in the same list. 

I'm willing to give your proposal a try, thanks for walking me through it. I'm still not sure I have an intuitive understanding of it, but perhaps attempting to implement it will clear things up.

**Comment by webbedspace** [comment]
One other thing I forgot to mention all this time: Nushell already has a notion of a suffixing `?` indicating "optional": `def a [foo?:int] {}` produces an optional argument using `?` to denote it. So, this symbology already exists in the language.

**Comment by rgwood** [comment]
I'm going to work on other things for a while. If anyone else wants to try implementing something like this, feel free.

**Comment by rgwood** [comment]
Closing this for now. I believe we still need something like this but I wasn't able to find a syntax that makes everyone happy.

Leon's suggestion might be the way to go but I wasn't able to wrap my head around it.

**Comment by fdncred** [comment]
If it were up to me, I'd land this PR even though I don't think the operators are "perfect". I'd also want to keep the `get -i` and `select -I` functionality until we are happy with how this works. 

If we're not deprecating anything with this PR and we're just adding new optional syntax, I'm not sure why we wouldn't land it and just keep iterating on it. I haven't looked at the details for a while so I may have forgotten something.

**Comment by rgwood** [comment]
I think enough people found the initial syntax unintuitive that there's clearly something wrong with it. I'd rather not spend time polishing and documenting this before we find a syntax people are happy with.

If anyone else want to run with this work, feel free; I've lost the motivation to work on it for now.

**Comment by webbedspace** [comment]
I want to take this branch and finish it myself… although I've been thinking a bit more about the semantics (as you might expect).

Something I've been pondering is what `?.` in other languages (C sharp, JS, TypeScript) was actually designed to do. In those languages, this syntax *only* works on explicit `null`/`undefined`. Why? Because it's meant to be used with well-defined structure shapes – things where a property name is explicitly marked `?`. They importantly do *not* work on holes (missing elements) because holes potentially denote a much more serious problem (that entirely the wrong class or object is being used here).

So, I've been wondering if using this syntax for holes at all is a good idea, given that it goes against this operator's "original" design intent. Maybe a second variant is needed, with a different symbol like `&.`, that only works on holes. But the fact that other languages generally don't have this (JS being an exception due to its hole-access semantics) gives me pause for whether this solves the problem correctly.

I do also think, in this light, that the `-i` flag is way too coarse-grained a solution for what it does, and really should be replaced with *something* more specific.

**Comment by fdncred** [comment]
For me, it's always been about the postfix operator meaning the item before it is optional. So, I think syntax should be closer to this.
Both of these have the same result.
```
[{foo: 'bar'} {}] | get foo?
[{foo: 'bar'} {}].foo?
```
```
╭───┬─────╮
│ 0 │ bar │
│ 1 │     │
╰───┴─────╯
```
Both of these have the same result.
```
[{foo: 'bar'} {}] | get foo
[{foo: 'bar'} {}].foo
```
```
Error: nu::shell::column_not_found (link)

  × Cannot find column
   ╭─[entry #19:1:1]
 1 │ [{foo: 'bar'} {}] | get foo
   · ────────┬────────   ─┬─
   ·         │            ╰── value originates here
   ·         ╰── cannot find column 'Empty cell'

```
And if you want to keep drilling down into nested structures it would be like:
`$var | get foo?.bar?.baz?` <-- foo, bar, and baz are optional
`$var | get foo?.bar.baz?` <-- foo and baz are optional but bar is required
`$var | get foo.bar.baz?` <-- foo and bar are required but baz is optional
the same with
`$var.foo?.bar?.baz?`
`$var.foo?.bar.baz?`
`$var.foo.bar.baz?`

**Comment by webbedspace** [comment]
So, like, do you really think it's OK to have this one cell path modifier produce `null` on both of these:
* Holes in structures (missing record key `{a:b}.c?`, list access-beyond-end `[].0?`, etc.)
* The container doesn't support that cell path format (`{a:b}.0?`, `$nothing.foo?`, `(2).0?`)

To me, these are different enough to merit consideration.

**Comment by rgwood** [comment]
> do you really think it's OK to have this one cell path modifier glyph produce `null` on both of these

I go back and forth on this. I like the simplicity of having 1 glyph in a known position suppress any error. But I can appreciate that those are meaningfully different types of errors, if we can find an ergonomic way to handle them separately let's do it.

Either way, I think we probably need short-circuiting behaviour (where we stop evaluating a cell path after a failure to access one optional cell path). I initially implemented this without short-circuiting and that was bad in retrospect; it's not very useful if `$var | get foo?.bar.baz?` succeeds on an optional access to `foo?` only to fail on an access to `bar` shortly thereafter.

**Comment by rgwood** [comment]
Good comment by @kubouch on Discord:

> Since we have `try`/`catch` now, I would lean towards `?` handling only "holes" in structures. `{a:b}.0?` and other examples listed in the second case should throw error IMO, which can be caught in `try`/`catch`.
>
> Yes, `try`/`catch` is more verbose than `-i` but at least it's obvious what is happening. (I was looking at some Fish scripts the other day and while it's "simple", the amount of flags on everything makes it impossible to figure out what is going on.)
>
> So maybe limiting the `?` to only "holes" might be a good starting point. It's also more focused so possibly less unintended side effects.

I think I agree with that.

**Comment by rgwood** [comment]
So, I think we want to try the following approach:

1. Postfix syntax: `?` goes after the path member e.g. `.foo?` instead of `?.foo`
2. `?` only suppresses errors from holes, incorrect types still generate an error

**Comment by daniel-vainsencher** [comment]
That sounds good to me. 

Cases other than optionality might be better addressed via a pattern-matching syntax, anything in that vein for nushell?

**Comment by rgwood** [comment]
New PR is up for postfix `?`: https://github.com/nushell/nushell/pull/8379

**Pull request closed without merging**